### PR TITLE
Reload chainTypes

### PR DESCRIPTION
**Pull request by jiqiang90**
- Fixed reload chainTypes when project changed, 

## Type of change

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Checklist

- [ ] I have tested locally
- [ ] I have performed a self review of my changes
- [ ] Updated any relevant documentation
- [ ] Linked to any relevant issues
- [ ] I have added tests relevant to my changes
- [ ] Any dependent changes have been merged and published in downstream modules
- [ ] My code is up to date with the base branch
- [ ] I have updated relevant changelogs. [We suggest using chan](https://github.com/geut/chan/tree/main/packages/chan)

### CODE REVIEW

File: packages/node/src/indexer/x-provider/http.ts [hunk 1]
```diff
@@ -54,7 +54,7 @@ export class HttpProvider implements ProviderInterface {
 
   readonly #stats: ProviderStats;
 
-  readonly #ctx: ReturnType<typeof context>;
+  readonly ctx: ReturnType<typeof context>;
```
Comment (stwiname) [inline packages/node/src/indexer/x-provider/http.ts:57]: Whats the reason for this change? It removes the privacy of the variable

File: packages/node/src/indexer/project.service.ts [hunk 2]
```diff
@@ -74,7 +74,8 @@ export class ProjectService extends BaseProjectService<
     return getTimestamp(block);
   }
 
-  protected onProjectChange(project: SubqueryProject): void | Promise<void> {
+  protected async onProjectChange(project: SubqueryProject): Promise<void> {
+    await this.apiService.reloadChainTypes(project.chainTypes);
     this.apiService.updateBlockFetching();
```
Comment (stwiname) [inline packages/node/src/indexer/project.service.ts:79]: Why not combine these into a single function?

File: packages/node-core/src/indexer/connectionPool.service.ts [hunk 3]
```diff
@@ -49,6 +49,7 @@ export class ConnectionPoolService<T extends IApiConnectionSpecific<any, any, an
 
   async onApplicationShutdown(): Promise<void> {
     await Promise.all(Object.values(this.allApi).map((api) => api?.apiDisconnect()));
+    this.allApi = {};
```
Comment (stwiname) [inline packages/node-core/src/indexer/connectionPool.service.ts:52]: What's the reason for this change?

File: packages/node/src/indexer/api.service.ts [hunk 4]
```diff
@@ -160,11 +160,24 @@ export class ApiService
...
+  }
+
+  async reloadChainTypes(chainTypes: unknown): Promise<void> {
+    // close any existing connections
+    await this.connectionPoolService.onApplicationShutdown();
+    logger.info(`Reload chainTypes`);
+    await this.load(chainTypes);
+  }
+
+  private async load(newChainTypes?: unknown): Promise<ApiService> {
```
Comment (stwiname) [inline packages/node/src/indexer/api.service.ts:173]: I think this could have a better function name.

File: packages/node/src/indexer/api.service.ts [hunk 5]
```diff
@@ -160,11 +160,24 @@ export class ApiService
   }
 
   async init(): Promise<ApiService> {
+    return this.load();
+  }
+
+  async reloadChainTypes(chainTypes: unknown): Promise<void> {
```
Comment (stwiname) [inline packages/node/src/indexer/api.service.ts:166]: Do we have to recreate all the api instances? This seems quite inefficient, creating new polkadot api instances is slow.

**Review by stwiname — COMMENTED** [review]
It would be good to separate the timezone changes to a separate PR to get a release out

File: packages/node/src/indexer/x-provider/http.ts (hunk 1 above)
Comment (jiqiang90) [reply packages/node/src/indexer/x-provider/http.ts:57]:
```
Cannot read private member #ctx from an object whose class did not declare it
TypeError: Cannot read private member #ctx from an object whose class did not declare it
    at Proxy.disconnect (/Users/jiatwork/ofn/subql/packages/node/src/indexer/x-provider/http.ts:122:16)
    at RpcCore.disconnect (/Users/jiatwork/ofn/subql/node_modules/@polkadot/rpc-core/cjs/bundle.js:103:30)

```

It throwing this error, I feel it should be safe to remove # here as it only accessed from polkadot

File: packages/node-core/src/indexer/connectionPool.service.ts (hunk 3 above)
Comment (jiqiang90) [reply packages/node-core/src/indexer/connectionPool.service.ts:52]: Trying to reset all existing connection here, but I think need a better place

File: packages/node/src/indexer/project.service.ts (hunk 2 above)
Comment (jiqiang90) [reply packages/node/src/indexer/project.service.ts:80]: `updateBlockFetching` also used in api service constructor, also make sense keep them independent.

File: packages/node/CHANGELOG.md [hunk 6]
```diff
@@ -6,6 +6,9 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## [Unreleased]
 
+### Fixed
+- Fixed api didn't reload new deployment chainTypes when project upgraded (#2505)
```
Comment (stwiname) [inline packages/node/CHANGELOG.md:10]:
```suggestion
- Fixed api not reloading new deployment chainTypes when project upgrades (#2505)
```

File: packages/node/src/indexer/apiPromise.connection.ts [hunk 7]
```diff
@@ -123,6 +131,14 @@ export class ApiPromiseConnection
     await this.unsafeApi.disconnect();
   }
 
+  async updateRegisteredChainTypes(chainTypes: RegisteredTypes): Promise<void> {
+    this.unsafeApi = await ApiPromiseConnection.createApiPromise(
```
Comment (stwiname) [inline packages/node/src/indexer/apiPromise.connection.ts:135]: This doesn't reuse the connection. If we have a websocket we don't want to create a new connection

File: packages/node/src/indexer/api.service.ts [hunk 8]
```diff
@@ -206,12 +206,16 @@ export class ApiService
...
   }
 
-  updateBlockFetching(): void {
-    const onlyEventHandlers = isOnlyEventHandlers(this.project);
+  async updateChainTypes(newChainTypes: unknown): Promise<void> {
+    const chainTypes = await updateChainTypesHasher(newChainTypes);
+    await this.connectionPoolService.updateChainTypes(chainTypes);
+  }
+
+  updateBlockFetching(project?: SubqueryProject): void {
```
Comment (stwiname) [inline packages/node/src/indexer/api.service.ts:217]: Theres no need to pass through the newChainTypes/project in these. `this.project` will be the updated project

File: packages/node/src/indexer/apiPromise.connection.ts (hunk 7 above)
Comment (jiqiang90) [reply packages/node/src/indexer/apiPromise.connection.ts:135]: image

New approach : Extract current ws/http provider from current apiPromise instance, and reconstructed a new apiPromise

File: packages/node/src/indexer/apiPromise.connection.ts [hunk 9]
```diff
@@ -123,6 +124,16 @@ export class ApiPromiseConnection
     await this.unsafeApi.disconnect();
   }
 
+  async updateChainTypes(chainTypes: RegisteredTypes): Promise<void> {
+    // Typeof Decorate<'promise' | 'rxjs'>, but we need to access this private method
+    const currentApiOptions = (this.unsafeApi as any)._options as ApiOptions;
+    const apiOption = {
+      ...currentApiOptions,
+      ...chainTypes,
+    };
```
Comment (stwiname) [inline packages/node/src/indexer/apiPromise.connection.ts:133]:
```suggestion
    const apiOption: ApiOptions = {
      ...currentApiOptions,
      chainTypes,
    };
```

**Review by stwiname — APPROVED** [review]

File: packages/node/src/indexer/apiPromise.connection.ts (hunk 9 above)
Comment (jiqiang90) [reply packages/node/src/indexer/apiPromise.connection.ts:133]: should be same line as https://github.com/subquery/subql/pull/2505/files#diff-c9f4c413e59e6f6047e10cd5abdc7e5aef91fd9d24a84732459c7d6acf395631R84
